Source/CarthageKit/Git.swift

```diff
@@ -294,83 +294,52 @@ public func checkoutRepositoryToDirectory(_ repositoryFileURL: URL, _ workingDir
 /// repository, but without any Git metadata.
 public func cloneSubmoduleInWorkingDirectory(_ submodule: Submodule, _ workingDirectoryURL: URL) -> SignalProducer<(), CarthageError> {
 	let submoduleDirectoryURL = workingDirectoryURL.appendingPathComponent(submodule.path, isDirectory: true)
+
+	func repositoryCheck<T>(attempt closure: () throws -> T, reasonForFailure: String) -> Result<T, CarthageError> {
+		do {
+			return .success(try closure())
+		} catch let error as NSError {
+			return .failure(
+				.repositoryCheckoutFailed(workingDirectoryURL: submoduleDirectoryURL, reason: reasonForFailure, underlyingError: error)
+			)
+		}
+	}
+
+	// swiftlint:disable switch_case_on_newline
 	let purgeGitDirectories = FileManager.default.reactive
 		.enumerator(at: submoduleDirectoryURL, includingPropertiesForKeys: [ .isDirectoryKey, .nameKey ], catchErrors: true)
-		.flatMap(.merge) { enumerator, url -> SignalProducer<(), CarthageError> in
-			var name: String?
-			do {
-				name = try url.resourceValues(forKeys: [ .nameKey ]).name
-			} catch let error as NSError {
-				return SignalProducer(
-					error: CarthageError.repositoryCheckoutFailed(
-						workingDirectoryURL: submoduleDirectoryURL,
-						reason: "could not enumerate name of descendant at \(url.path)",
-						underlyingError: error
-					)
-				)
+		.attemptMap { enumerator, url -> Result<(), CarthageError> in
+			switch repositoryCheck(attempt: {
+				try url.resourceValues(forKeys: [ .nameKey ]).name
+			}, reasonForFailure: "could not enumerate name of descendant at \(url.path)") {
+			case .failure(let error): return .failure(error)
```

Comment:
I think this could be structured to use `Result.flatMap` instead of returning the error like this. That would move the `name != ".git"` logic into the block below.